presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java

```diff
@@ -0,0 +1,398 @@
+            Duration completedQueryExpirationTimeout,
+            Duration nodeStatusTimeout,
+            Duration memoryPoolInfoRefreshDuration,
+            boolean isReservedPoolEnabled,
+            ScheduledExecutorService scheduledExecutorService)
+    {
+        this.internalNodeManager = requireNonNull(internalNodeManager, "internalNodeManager is null");
+        checkArgument(maxCompletedQueries > 0, "maxCompletedQueries must be > 0, was %s", maxCompletedQueries);
+        this.maxCompletedQueries = maxCompletedQueries;
+        this.queryExpirationTimeout = requireNonNull(queryExpirationTimeout, "queryExpirationTimeout is null");
+        this.completedQueryExpirationTimeout = requireNonNull(completedQueryExpirationTimeout, "completedQueryExpirationTimeout is null");
+        // Memoized suppliers take in a time unit > 0
+        requireNonNull(memoryPoolInfoRefreshDuration, "memoryPoolInfoRefreshDuration is null");
+        if (memoryPoolInfoRefreshDuration.toMillis() > 0) {
+            this.clusterMemoryPoolInfosSupplier = Suppliers.memoizeWithExpiration(this::getClusterMemoryPoolInfoInternal, memoryPoolInfoRefreshDuration.toMillis(), MILLISECONDS);
+        }
+        else {
+            this.clusterMemoryPoolInfosSupplier = this::getClusterMemoryPoolInfoInternal;
+        }
+        this.isReservedPoolEnabled = isReservedPoolEnabled;
+
+        requireNonNull(scheduledExecutorService, "scheduledExecutorService is null");
+        scheduledExecutorService.scheduleAtFixedRate(() -> {
+            for (CoordinatorQueriesState coordinatorQueriesState : ImmutableList.copyOf(nodeQueryStates.values())) {
+                coordinatorQueriesState.purgeExpiredQueries();
+            }
+        }, 100, 100, MILLISECONDS);
+
+        scheduledExecutorService.scheduleAtFixedRate(() -> {
+            for (Map.Entry<String, InternalNodeState> nodeEntry : ImmutableList.copyOf(nodeStatuses.entrySet())) {
+                if ((System.currentTimeMillis() - nodeEntry.getValue().getLastHeartbeatInMillis()) > nodeStatusTimeout.toMillis()) {
+                    nodeStatuses.remove(nodeEntry.getKey());
+                }
+            }
+        }, 100, 100, MILLISECONDS);
+    }
+
+    public void registerQueryHeartbeat(String nodeId, BasicQueryInfo basicQueryInfo)
+    {
+        requireNonNull(nodeId, "nodeId is null");
+        requireNonNull(basicQueryInfo, "basicQueryInfo is null");
+        checkArgument(
+                internalNodeManager.getCoordinators().stream().anyMatch(i -> nodeId.equals(i.getNodeIdentifier())),
+                "%s is not a coordinator",
+                nodeId);
+        CoordinatorQueriesState state = nodeQueryStates.computeIfAbsent(nodeId, identifier -> new CoordinatorQueriesState(
+                nodeId,
+                maxCompletedQueries,
+                queryExpirationTimeout.toMillis(),
+                completedQueryExpirationTimeout.toMillis()));
+        state.addOrUpdateQuery(basicQueryInfo);
```

Comment:
This method will be invoked directly from IO thread. If we have a relatively long-running `purgeExpiredQueries`, we might choke the IO thread pool. Given we only need the ordering or completed queries, maybe make `activeQueries` a concurrent hash map and `completedQueries` a concurrent queue. Of course, in some place, we will be careful updating both data structures. For example in `addOrUpdateQuery`, we need to have:
```java
if (activeQueries.remove(queryId) != null) {
    completedQueries.put(query.getQueryId(), query);
}

Reply:
The chief difficulty around this is that `completedQueries` may contain duplicates, and it would be expensive to de-duplicate it in `getAllQueries`.  The most straightforward thing may be to just move this out of the IO thread.